indexer/PresetsFile.js

```diff
@@ -4,47 +4,109 @@ const crypto = require('crypto');
 
 class PresetsFile
 {
-    constructor(fullPath, presetsFileEncoding)
+    constructor(fullPath, presetsFileEncoding, presetsFileMetadata, errors)
     {
         this.fullPath = fullPath;
-        this.title = "";
-        this.version = "";
-        this.description = [];
         this.hash = "";
-
-        this._remainingHeaders = ["title", "version"];
+        this._presetsFileMetadata = presetsFileMetadata;
+        this._errors = errors;
 
         const binaryFileContent = fs.readFileSync(this.fullPath);
         let sum = crypto.createHash('sha512');
         sum.update(binaryFileContent);
         this.hash = sum.digest('hex');
 
+        this._processLines(binaryFileContent, presetsFileEncoding);
+        this._checkProperties();
+
+        delete this._presetsFileMetadata;
+        delete this._errors;
+    }
+
+    _checkProperties()
+    {
+        for (const [property, value] of Object.entries(this._presetsFileMetadata)) {
+            if (!value.optional) {
+                if (this._isEmptyProperty(this[property]))
+                {
+                    this._addError(`missing or empty ${property}`);
+                }
+            }
+        }
+    }
+
+    _isEmptyProperty(property)
+    {
+        return (property === undefined || property.length === 0);
+    }
+
+    _processLines(binaryFileContent, presetsFileEncoding)
+    {
         const fileContent = binaryFileContent.toString(presetsFileEncoding);
         const lines = fileContent.split('\n');
 
+        this._currentLine = 1;
+
         for (let line of lines) {
             line = line.trim();
             if (line.startsWith("#")) {
-                this._processHeaderLine(line);
-            } else {
-                break;
+                this._processCommentLine(line);
+            }
+
+            this._currentLine++;
+        }
+
+        delete this._currentLine;
+    }
+
+    _processCommentLine(line)
+    {
+        line = line.slice(1).trim(); // (# Title: foo) -> (Title: foo)
+        const lowCaseLine = line.toLowerCase();
+
+        for (const [property, value] of Object.entries(this._presetsFileMetadata)) {
+            const lineBeginning = `${property.toLowerCase()}:`; // "Title:"
+
+            if (lowCaseLine.startsWith(lineBeginning)) {
+                line = line.slice(lineBeginning.length).trim(); // (Title: foo) -> (foo)
+                this._processProperty(property, line);
             }
         }
+    }
 
-        delete this._remainingHeaders;
+    _processProperty(property, line)
+    {
+        if (Array.isArray(this._presetsFileMetadata[property].value)) {
```

Comment:
Interesting way of doing things. :-)

Reply:
ha! I just wanted so that it was easy and transparent how to add/remove presets metadata properties and configure them if a property is just a string or array :)

Reply:
We'll see how well this works once we want to add support for other data types. :wink: